File: apps/mp/src/utils/permissions.js
/** H5 权限与菜单（与 Web / SRS §2.5 一致） */
export function hasPermission(permissions, code) {
  return (permissions || []).includes(code)
}

export function hasAnyPermission(permissions, codes) {
  const set = new Set(permissions || [])
  return codes.some((c) => set.has(c))
}

export function isPermDisabled(permissions, item) {
  return item.requires && !permissions.includes(item.requires)
}

export const PERMISSION_GROUPS = [
  {
    label: '工作台',
    rows: [
      {
        type: 'scope',
        menu: '工作台',
        items: [
          { code: 'dashboard.view', label: '访问' },
          { code: 'dashboard.view_all', label: '全公司统计', requires: 'dashboard.view' },
        ],
      },
    ],
  },
  {
    label: '数据看板',
    rows: [
      {
        type: 'scope',
        menu: '数据看板',
        items: [
          { code: 'analytics.view', label: '访问' },
          { code: 'analytics.view_all', label: '全公司统计', requires: 'analytics.view' },
        ],
      },
    ],
  },
  {
    label: '内容',
    rows: [
      { type: 'inline', items: [{ code: 'content.create', label: '营销创作' }] },
      {
        type: 'scope',
        menu: '内容库',
        items: [
          { code: 'content.list_own', label: '本人' },
          { code: 'content.list_all', label: '全公司', requires: 'content.list_own' },
        ],
      },
      {
        type: 'scope',
        menu: '查看文章',
        items: [
          { code: 'content.view_own', label: '本人' },
          { code: 'content.view_all', label: '任意', requires: 'content.view_own' },
        ],
      },
      {
        type: 'inline',
        items: [
          { code: 'content.edit', label: '编辑内容' },
          { code: 'content.delete', label: '删除内容' },
          { code: 'content.export', label: '导出' },
          { code: 'content.schedule', label: '发布日历' },
          { code: 'content.publish', label: '发布公众号' },
        ],
      },
    ],
  },
  {
    label: '知识库',
    rows: [
      {
        type: 'scope',
        menu: '知识库',
        items: [
          { code: 'knowledge.view', label: '查看' },
          { code: 'knowledge.manage', label: '管理', requires: 'knowledge.view' },
        ],
      },
    ],
  },
  {
    label: '设置',
    rows: [
      {
        type: 'inline',
        items: [
          { code: 'preference.manage', label: '我的偏好' },
          { code: 'brand.manage', label: '品牌设置' },
          { code: 'wechat.manage', label: '公众号绑定' },
          { code: 'llm.manage', label: 'AI 模型' },
          { code: 'tenant.manage', label: '企业信息' },
        ],
      },
    ],
  },
  {
    label: '团队',
    rows: [
      {
        type: 'scope',
        menu: '成员',
        items: [
          { code: 'team.member.view', label: '查看' },
          { code: 'team.member.manage', label: '管理', requires: 'team.member.view' },
        ],
      },
      { type: 'inline', items: [{ code: 'team.role.manage', label: '管理角色', requires: 'team.member.view' }] },
    ],
  },
  {
    label: 'CRM — 线索',
    rows: [
      {
        type: 'scope',
        menu: '线索列表',
        items: [
          { code: 'crm.lead.list_own', label: '本人' },
          { code: 'crm.lead.list_team', label: '下级', requires: 'crm.lead.list_own' },
          { code: 'crm.lead.list_territory', label: '地区', requires: 'crm.lead.list_own' },
          { code: 'crm.lead.list_all', label: '全公司', requires: 'crm.lead.list_own' },
        ],
      },
      {
        type: 'inline',
        items: [
          { code: 'crm.lead.view', label: '查看', requires: 'crm.lead.list_own' },
          { code: 'crm.lead.create', label: '新建', requires: 'crm.lead.view' },
          { code: 'crm.lead.edit', label: '编辑', requires: 'crm.lead.view' },
          { code: 'crm.lead.assign', label: '分配负责人', requires: 'crm.lead.edit' },
          { code: 'crm.lead.convert', label: '转化客户', requires: 'crm.lead.edit' },
          { code: 'crm.lead.delete', label: '删除', requires: 'crm.lead.edit' },
          { code: 'crm.lead.import', label: '导入', requires: 'crm.lead.create' },
        ],
      },
    ],
  },
  {
    label: 'CRM — 客户',
    rows: [
      {
        type: 'scope',
        menu: '客户列表',
        items: [
          { code: 'crm.customer.list_own', label: '本人' },
          { code: 'crm.customer.list_team', label: '下级', requires: 'crm.customer.list_own' },
          { code: 'crm.customer.list_territory', label: '地区', requires: 'crm.customer.list_own' },
          { code: 'crm.customer.list_all', label: '全公司', requires: 'crm.customer.list_own' },
        ],
      },
      {
        type: 'inline',
        items: [
          { code: 'crm.customer.view', label: '查看', requires: 'crm.customer.list_own' },
          { code: 'crm.customer.create', label: '新建', requires: 'crm.customer.view' },
          { code: 'crm.customer.edit', label: '编辑', requires: 'crm.customer.view' },
          { code: 'crm.customer.assign', label: '分配负责人', requires: 'crm.customer.edit' },
          { code: 'crm.customer.delete', label: '删除', requires: 'crm.customer.edit' },
          { code: 'crm.customer.import', label: '导入', requires: 'crm.customer.create' },
        ],
      },
    ],
  },
  {
    label: 'CRM — 任务',
    rows: [
      {
        type: 'scope',
        menu: '任务列表',
        items: [
          { code: 'crm.task.list_own', label: '本人' },
          { code: 'crm.task.list_team', label: '下级', requires: 'crm.task.list_own' },
          { code: 'crm.task.list_territory', label: '地区', requires: 'crm.task.list_own' },
          { code: 'crm.task.list_all', label: '全公司', requires: 'crm.task.list_own' },
        ],
      },
      {
        type: 'inline',
        items: [
          { code: 'crm.task.create', label: '新建', requires: 'crm.task.list_own' },
          { code: 'crm.task.edit', label: '编辑', requires: 'crm.task.list_own' },
          { code: 'crm.task.assign', label: '指派执行人', requires: 'crm.task.edit' },
          { code: 'crm.task.delete', label: '删除', requires: 'crm.task.edit' },
        ],
      },
    ],
  },
  {
    label: 'CRM — 营销活动',
    rows: [
      {
        type: 'scope',
        menu: '活动列表',
        items: [
          { code: 'crm.campaign.list_own', label: '本人' },
          { code: 'crm.campaign.list_team', label: '下级', requires: 'crm.campaign.list_own' },
          { code: 'crm.campaign.list_territory', label: '地区', requires: 'crm.campaign.list_own' },
          { code: 'crm.campaign.list_all', label: '全公司', requires: 'crm.campaign.list_own' },
        ],
      },
      {
        type: 'inline',
        items: [
          { code: 'crm.campaign.view', label: '查看', requires: 'crm.campaign.list_own' },
          { code: 'crm.campaign.create', label: '新建', requires: 'crm.campaign.view' },
          { code: 'crm.campaign.edit', label: '编辑', requires: 'crm.campaign.view' },
          { code: 'crm.campaign.manage', label: '管理状态', requires: 'crm.campaign.edit' },
          { code: 'crm.campaign.delete', label: '删除', requires: 'crm.campaign.edit' },
        ],
      },
    ],
  },
  {
    label: 'CRM — 组织与表单',
    rows: [
      {
        type: 'inline',
        items: [
          { code: 'crm.org.manage', label: '销售组织' },
          { code: 'crm.schema.manage', label: '表单字段管理' },
          { code: 'crm.activity.create', label: '写跟进记录' },
        ],
      },
      {
        type: 'scope',
        menu: '列表视图',
        items: [
          { code: 'crm.view.save_own', label: '保存私有' },
          { code: 'crm.view.manage_public', label: '管理公开', requires: 'crm.view.save_own' },
        ],
      },
    ],
  },
]

export const CRM_MENU = [
  {
    title: '线索',
    desc: '跟进销售线索',
    url: '/pages/crm/leads',
    icon: '👤',
    iconColor: '#1677ff',
    permissionAny: ['crm.lead.list_own', 'crm.lead.list_team', 'crm.lead.list_territory', 'crm.lead.list_all'],
  },
  {
    title: '客户',
    desc: '管理成交客户',
    url: '/pages/crm/customers',
    icon: '👥',
    iconColor: '#52c41a',
    permissionAny: ['crm.customer.list_own', 'crm.customer.list_team', 'crm.customer.list_territory', 'crm.customer.list_all'],
  },
  {
    title: '商机',
    desc: '跟进销售机会',
    url: '/pages/crm/deals',
    icon: '📈',
    iconColor: '#fa8c16',
    permissionAny: ['crm.deal.list_own', 'crm.deal.list_team', 'crm.deal.list_territory', 'crm.deal.list_all'],
  },
  {
    title: '订单',
    desc: '查看与管理订单',
    url: '/pages/crm/orders',
    icon: '💳',
    iconColor: '#0958d9',
    permissionAny: ['crm.order.list_own', 'crm.order.list_team', 'crm.order.list_territory', 'crm.order.list_all'],
  },
  {
    title: '报价',
    desc: '客户报价单',
    url: '/pages/crm/quotes',
    icon: '📄',
    iconColor: '#1677ff',
    permissionAny: ['crm.quote.list_own', 'crm.quote.list_all'],
  },
  {
    title: '合同',
    desc: '销售合同',
    url: '/pages/crm/contracts',
    icon: '📋',
    iconColor: '#722ed1',
    permissionAny: ['crm.contract.list_own', 'crm.contract.list_all'],
  },
  {
    title: '回款',
    desc: '回款登记与确认',
    url: '/pages/crm/payments',
    icon: '💰',
    iconColor: '#ff4d4f',
    permissionAny: [
      'crm.payment.list_own',
      'crm.payment.list_team',
      'crm.payment.list_territory',
      'crm.payment.list_all',
    ],
  },
  {
    title: '我的任务',
    desc: '待办与回访',
    url: '/pages/crm/tasks',
    icon: '📝',
    iconColor: '#1677ff',
    permissionAny: ['crm.task.list_own', 'crm.task.list_team', 'crm.task.list_territory', 'crm.task.list_all'],
  },
  {
    title: '营销活动',
    desc: '查看活动与线索',
    url: '/pages/crm/campaigns',
    icon: '🎯',
    iconColor: '#fa8c16',
    permissionAny: [
      'crm.campaign.list_own',
      'crm.campaign.list_team',
      'crm.campaign.list_territory',
      'crm.campaign.list_all',
    ],
  },
]

export function filterMenuByPermission(items, permissions) {
  return items.filter((item) => {
    if (item.permissionAny) return hasAnyPermission(permissions, item.permissionAny)
    return hasPermission(permissions, item.permission)
  })
}

export function hasAnyCrmListPermission(permissions) {
  return CRM_MENU.some((item) => hasAnyPermission(permissions, item.permissionAny))
}

export const MINE_MENU = [
  ...CRM_MENU,
  { title: '企业信息', url: '/pages/settings/tenant', permission: 'tenant.manage' },
  { title: '角色与成员', url: '/pages/settings/team', permissionAny: ['team.member.view', 'team.role.manage'] },
  { title: '我的偏好', url: '/pages/settings/preference', permission: 'preference.manage' },
  { title: 'AI 记忆', url: '/pages/settings/memory', permission: 'content.create' },
  { title: '公众号绑定', url: '/pages/wechat/wechat', permission: 'wechat.manage' },
]

File: apps/api/app/permissions.py
"""企业功能权限 Catalog（与 docs/00-总览/需求规格.md §2.5、§2.8 一致）。

内容获客商城 Phase1 权限见 docs/01-PRD/21-内容获客商城-phase1/05-角色权限.html#catalog
"""

from __future__ import annotations

_BASE_PERMISSIONS: tuple[str, ...] = (
    # 创作与内容
    "content.create",
    "content.list_own",
    "content.view_own",
    "content.list_all",
    "content.view_all",
    "content.edit",
    "content.delete",
    "content.export",
    "content.schedule",
    "content.publish",
    # 知识库
    "knowledge.view",
    "knowledge.manage",
    # 设置
    "preference.manage",
    "brand.manage",
    "wechat.manage",
    "llm.manage",
    "tenant.manage",
    # 工作台与数据看板
    "dashboard.view",
    "dashboard.view_all",
    "analytics.view",
    "analytics.view_all",
    # 团队
    "team.member.view",
    "team.member.manage",
    "team.role.manage",
)

CRM_PERMISSIONS: tuple[str, ...] = (
    # 线索
    "crm.lead.list_own",
    "crm.lead.list_team",
    "crm.lead.list_territory",
    "crm.lead.list_all",
    "crm.lead.view",
    "crm.lead.create",
    "crm.lead.edit",
    "crm.lead.assign",
    "crm.lead.convert",
    "crm.lead.delete",
    # 客户
    "crm.customer.list_own",
    "crm.customer.list_team",
    "crm.customer.list_territory",
    "crm.customer.list_all",
    "crm.customer.view",
    "crm.customer.create",
    "crm.customer.edit",
    "crm.customer.assign",
    "crm.customer.delete",
    # 任务
    "crm.task.list_own",
    "crm.task.list_team",
    "crm.task.list_territory",
    "crm.task.list_all",
    "crm.task.create",
    "crm.task.edit",
    "crm.task.assign",
    "crm.task.delete",
    # 营销活动
    "crm.campaign.list_own",
    "crm.campaign.list_team",
    "crm.campaign.list_territory",
    "crm.campaign.list_all",
    "crm.campaign.view",
    "crm.campaign.create",
    "crm.campaign.edit",
    "crm.campaign.manage",
    "crm.campaign.delete",
    # 组织、表单、跟进、视图、导入
    "crm.org.manage",
    "crm.schema.manage",
    "crm.activity.create",
    "crm.view.save_own",
    "crm.view.manage_public",
    "crm.lead.import",
    "crm.customer.import",
    # v0.7 商机与管道
    "crm.deal.list_own",
    "crm.deal.list_team",
    "crm.deal.list_territory",
    "crm.deal.list_all",
    "crm.deal.view",
    "crm.deal.create",
    "crm.deal.edit",
    "crm.deal.assign",
    "crm.deal.convert",
    "crm.deal.close",
    "crm.deal.reopen",
    "crm.deal.delete",
    "crm.pipeline.manage",
    # v0.7 产品 / v1.4 导入
    "crm.product.manage",
    "crm.product.import",
    # v0.7 报价
    "crm.quote.list_own",
    "crm.quote.list_team",
    "crm.quote.list_all",
    "crm.quote.view",
    "crm.quote.create",
    "crm.quote.edit",
    "crm.quote.send",
    "crm.quote.accept",
    "crm.quote.delete",
    # v0.7 合同
    "crm.contract.list_own",
    "crm.contract.list_team",
    "crm.contract.list_all",
    "crm.contract.view",
    "crm.contract.create",
    "crm.contract.edit",
    "crm.contract.sign",
    "crm.contract.approve",
    "crm.contract.delete",
    # v0.7 订单
    "crm.order.list_own",
    "crm.order.list_team",
    "crm.order.list_territory",
    "crm.order.list_all",
    "crm.order.view",
    "crm.order.create",
    "crm.order.edit",
    "crm.order.assign",
    "crm.order.place",
    "crm.order.approve",
    "crm.order.convert",
    "crm.order.delete",
    # v0.7 收款
    "crm.payment.list_own",
    "crm.payment.list_team",
    "crm.payment.list_territory",
    "crm.payment.list_all",
    "crm.payment.view",
    "crm.payment.create",
    "crm.payment.edit",
    "crm.payment.confirm",
    "crm.payment.reverse",
    "crm.payment.delete",
)

# 内容获客商城 · 平台运营端（跨租户，不参与 Membership）
PLATFORM_SHOP_PERMISSIONS: tuple[str, ...] = (
    "platform.shop.analytics",
    "platform.shop.approve",
    "platform.shop.channel",
    "platform.shop.fee.manage",
    "platform.shop.merchant.assign",
    "platform.shop.merchant.list_all",
    "platform.shop.merchant.list_assigned",
    "platform.shop.merchant.manage",
    "platform.shop.merchant.read",
    "platform.shop.merchant.tag",
    "platform.shop.merchant.tag.manage",
    "platform.shop.moderate",
    "platform.shop.onboarding.initiate",
    "platform.shop.plan.manage",
    "platform.shop.product.force_off",
    "platform.shop.product.review",
    "platform.shop.settlement",
    "platform.shop.subscription.manage",
    "platform.shop.subscription.read",
)

# 内容获客商城 · 商家端（租户内 shop.*）
SHOP_MERCHANT_PERMISSIONS: tuple[str, ...] = (
    "shop.analytics.read",
    "shop.buyer.list_all",
    "shop.buyer.view",
    "shop.channel.map",
    "shop.channel.read",
    "shop.channel.write",
    "shop.content.read",
    "shop.content.write",
    "shop.entitlement.list_all",
    "shop.entitlement.revoke",
    "shop.entitlement.view",
    "shop.invoice.list_all",
    "shop.invoice.process",
    "shop.invoice.view",
    "shop.order.close",
    "shop.order.export",
    "shop.order.list_all",
    "shop.order.list_own",
    "shop.order.refund",
    "shop.order.resend_notify",
    "shop.order.view",
    "shop.product.delete",
    "shop.product.publish",
    "shop.product.read",
    "shop.product.submit_review",
    "shop.product.write",
    "shop.redemption.execute",
    "shop.redemption.list_all",
    "shop.redemption.list_own",
    "shop.redemption.read",
    "shop.role.manage",
    "shop.settings.read",
    "shop.settings.write",
    "shop.store.manage",
    "shop.store.settings.read",
    "shop.store.settings.write",
    "shop.subscription.usage.read",
)

# 全部权限 code（租户 Membership；不含 platform.shop.*）
ALL_PERMISSIONS: tuple[str, ...] = _BASE_PERMISSIONS + CRM_PERMISSIONS + SHOP_MERCHANT_PERMISSIONS

# editor 内置角色默认权限（§2.5 带 ✅）
EDITOR_DEFAULT_PERMISSIONS: frozenset[str] = frozenset(
    {
        "content.create",
        "content.list_own",
        "content.view_own",
        "preference.manage",
        "dashboard.view",
        "analytics.view",
    }
)

SALES_DEFAULT_PERMISSIONS: frozenset[str] = frozenset(
    {
        "preference.manage",
        "dashboard.view",
        "analytics.view",
        "team.member.view",
        "crm.lead.list_own",
        "crm.lead.view",
        "crm.lead.create",
        "crm.lead.edit",
        "crm.lead.convert",
        "crm.customer.list_own",
        "crm.customer.view",
        "crm.customer.create",
        "crm.customer.edit",
        "crm.activity.create",
        "crm.task.list_own",
        "crm.task.create",
        "crm.task.edit",
        "crm.view.save_own",
        "crm.lead.import",
        "crm.customer.import",
        # v0.7 商机
        "crm.deal.list_own",
        "crm.deal.view",
        "crm.deal.create",
        "crm.deal.edit",
        "crm.deal.convert",
        "crm.deal.close",
        # v0.7 报价
        "crm.quote.list_own",
        "crm.quote.view",
        "crm.quote.create",
        "crm.quote.edit",
        # v0.7 合同
        "crm.contract.list_own",
        "crm.contract.view",
        # v0.7 订单
        "crm.order.list_own",
        "crm.order.view",
        "crm.order.create",
        "crm.order.edit",
        "crm.order.place",
        "crm.order.convert",
        # v0.7 收款
        "crm.payment.list_own",
        "crm.payment.view",
        "crm.payment.create",
        "crm.payment.edit",
    }
)

SALES_MANAGER_DEFAULT_PERMISSIONS: frozenset[str] = SALES_DEFAULT_PERMISSIONS | frozenset(
    {
        "crm.lead.list_team",
        "crm.lead.list_territory",
        "crm.customer.list_team",
        "crm.customer.list_territory",
        "crm.task.list_team",
        "crm.task.list_territory",
        "crm.lead.assign",
        "crm.customer.assign",
        "crm.task.assign",
        "crm.lead.delete",
        "crm.customer.delete",
        "crm.view.manage_public",
        "dashboard.view_all",
        "analytics.view_all",
        "team.member.view",
        # v0.7 商机
        "crm.deal.list_team",
        "crm.deal.list_territory",
        "crm.deal.assign",
        "crm.deal.reopen",
        "crm.deal.delete",
        # v0.7 报价（与商机一致：经理看团队，admin 全公司）
        "crm.quote.list_team",
        "crm.quote.send",
        "crm.quote.accept",
        "crm.quote.delete",
        # v0.7 合同
        "crm.contract.list_team",
        "crm.contract.create",
        "crm.contract.edit",
        "crm.contract.sign",
        "crm.contract.approve",
        "crm.contract.delete",
        # v0.7 订单
        "crm.order.list_team",
        "crm.order.list_territory",
        "crm.order.assign",
        "crm.order.approve",
        "crm.order.delete",
        # v0.7 收款
        "crm.payment.list_team",
        "crm.payment.list_territory",
        "crm.payment.confirm",
        "crm.payment.reverse",
        "crm.payment.delete",
    }
)

MARKETING_DEFAULT_PERMISSIONS: frozenset[str] = frozenset(
    {
        "preference.manage",
        "dashboard.view",
        "analytics.view",
        "content.create",
        "content.list_own",
        "content.view_own",
        "content.export",
        "content.schedule",
        "knowledge.view",
        "crm.campaign.list_own",
        "crm.campaign.view",
        "crm.campaign.create",
        "crm.campaign.edit",
        "crm.campaign.manage",
        "crm.campaign.list_all",
        "crm.lead.list_own",
        "crm.lead.view",
        "crm.lead.create",
        "crm.lead.edit",
        "crm.lead.convert",
        "crm.lead.list_all",
        "crm.customer.list_own",
        "crm.customer.view",
        "crm.view.save_own",
        # v0.7 商机（只读 + 新建）
        "crm.deal.list_own",
        "crm.deal.view",
        "crm.deal.create",
        # v0.7 报价（只读）
        "crm.quote.list_own",
        "crm.quote.view",
        # v0.7 合同（只读）
        "crm.contract.list_own",
        "crm.contract.view",
        # v0.7 订单（只读）
        "crm.order.list_own",
        "crm.order.view",
    }
)

SYSTEM_ROLE_CODES: frozenset[str] = frozenset(
    {"admin", "editor", "sales", "sales_manager", "marketing"}
)

SYSTEM_ROLE_ADMIN = "admin"
SYSTEM_ROLE_EDITOR = "editor"
SYSTEM_ROLE_SALES = "sales"
SYSTEM_ROLE_SALES_MANAGER = "sales_manager"
SYSTEM_ROLE_MARKETING = "marketing"

# 商城商家端内置角色（Phase1 种子化，见 05-角色权限#roles）
SYSTEM_ROLE_SHOP_ADMIN = "shop_admin"
SYSTEM_ROLE_SHOP_CONTENT = "shop_content"
SYSTEM_ROLE_SHOP_SUPPORT = "shop_support"
SYSTEM_ROLE_SHOP_CLERK = "shop_clerk"

SHOP_BUILTIN_ROLE_CODES: frozenset[str] = frozenset(
    {
        SYSTEM_ROLE_SHOP_ADMIN,
        SYSTEM_ROLE_SHOP_CONTENT,
        SYSTEM_ROLE_SHOP_SUPPORT,
        SYSTEM_ROLE_SHOP_CLERK,
    }
)

SHOP_ADMIN_DEFAULT_PERMISSIONS: frozenset[str] = frozenset(SHOP_MERCHANT_PERMISSIONS) - frozenset(
    {
        "shop.role.manage",
        "shop.channel.write",  # 租户级公域对接（选链路/路径/回调验通）仅企业管理员
    }
)

SHOP_CONTENT_DEFAULT_PERMISSIONS: frozenset[str] = frozenset(
    {
        "shop.analytics.read",
        "shop.content.read",
        "shop.content.write",
        "shop.product.delete",
        "shop.product.publish",
        "shop.product.read",
        "shop.product.submit_review",
        "shop.product.write",
        "shop.subscription.usage.read",
    }
)

SHOP_SUPPORT_DEFAULT_PERMISSIONS: frozenset[str] = frozenset(
    {
        "shop.analytics.read",
        "shop.buyer.list_all",
        "shop.buyer.view",
        "shop.content.read",
        "shop.entitlement.list_all",
        "shop.entitlement.revoke",
        "shop.entitlement.view",
        "shop.invoice.list_all",
        "shop.invoice.process",
        "shop.invoice.view",
        "shop.order.export",
        "shop.order.list_all",
        "shop.order.refund",
        "shop.order.resend_notify",
        "shop.order.view",
        "shop.product.read",
        "shop.redemption.read",
        "shop.subscription.usage.read",
    }
)

SHOP_CLERK_DEFAULT_PERMISSIONS: frozenset[str] = frozenset(
    {
        "shop.redemption.execute",
        "shop.redemption.list_own",
        "shop.redemption.read",
    }
)


def membership_is_shop_clerk(membership) -> bool:
    """对照 #a08-clerk：店员壳判定。"""
    role = getattr(membership, "role", None) if membership is not None else None
    return getattr(role, "code", None) == SYSTEM_ROLE_SHOP_CLERK

SHOP_BUILTIN_ROLE_DEFAULT_PERMISSIONS: dict[str, frozenset[str]] = {
    SYSTEM_ROLE_SHOP_ADMIN: SHOP_ADMIN_DEFAULT_PERMISSIONS,
    SYSTEM_ROLE_SHOP_CONTENT: SHOP_CONTENT_DEFAULT_PERMISSIONS,
    SYSTEM_ROLE_SHOP_SUPPORT: SHOP_SUPPORT_DEFAULT_PERMISSIONS,
    SYSTEM_ROLE_SHOP_CLERK: SHOP_CLERK_DEFAULT_PERMISSIONS,
}

# 平台商城子角色（挂接 platform_admin 账号，Phase1 可选模板）
PLATFORM_SHOP_ROLE_OPS = "platform_shop_ops"
PLATFORM_SHOP_ROLE_CS = "platform_shop_cs"
PLATFORM_SHOP_ROLE_FINANCE = "platform_shop_finance"

PLATFORM_SHOP_ROLE_CODES: frozenset[str] = frozenset(
    {
        PLATFORM_SHOP_ROLE_OPS,
        PLATFORM_SHOP_ROLE_CS,
        PLATFORM_SHOP_ROLE_FINANCE,
    }
)

PLATFORM_SHOP_OPS_DEFAULT_PERMISSIONS: frozenset[str] = frozenset(
    {
        "platform.shop.analytics",
        "platform.shop.approve",
        "platform.shop.merchant.assign",
        "platform.shop.merchant.list_all",
        "platform.shop.merchant.manage",
        "platform.shop.merchant.read",
        "platform.shop.merchant.tag",
        "platform.shop.merchant.tag.manage",
        "platform.shop.moderate",
        "platform.shop.plan.manage",
        "platform.shop.product.force_off",
        "platform.shop.product.review",
        "platform.shop.subscription.manage",
        "platform.shop.subscription.read",
    }
)

PLATFORM_SHOP_CS_DEFAULT_PERMISSIONS: frozenset[str] = frozenset(
    {
        "platform.shop.analytics",
        "platform.shop.merchant.list_assigned",
        "platform.shop.merchant.read",
        "platform.shop.merchant.tag",
        "platform.shop.onboarding.initiate",
        "platform.shop.subscription.read",
    }
)

PLATFORM_SHOP_FINANCE_DEFAULT_PERMISSIONS: frozenset[str] = frozenset(
    {
        "platform.shop.analytics",
        "platform.shop.fee.manage",
        "platform.shop.settlement",
        "platform.shop.subscription.read",
    }
)

PLATFORM_SHOP_ROLE_DEFAULT_PERMISSIONS: dict[str, frozenset[str]] = {
    PLATFORM_SHOP_ROLE_OPS: PLATFORM_SHOP_OPS_DEFAULT_PERMISSIONS,
    PLATFORM_SHOP_ROLE_CS: PLATFORM_SHOP_CS_DEFAULT_PERMISSIONS,
    PLATFORM_SHOP_ROLE_FINANCE: PLATFORM_SHOP_FINANCE_DEFAULT_PERMISSIONS,
}

PLATFORM_SHOP_PERMISSION_LABELS: dict[str, str] = {
    "platform.shop.analytics": "经营看板只读",
    "platform.shop.approve": "入驻审核",
    "platform.shop.channel": "公域渠道与支付进件",
    "platform.shop.fee.manage": "类目与费率",
    "platform.shop.merchant.assign": "分配管家",
    "platform.shop.merchant.list_all": "全站商家数据范围",
    "platform.shop.merchant.list_assigned": "仅看分配给自己的商家",
    "platform.shop.merchant.manage": "暂停/恢复",
    "platform.shop.merchant.read": "商家列表/详情 + 跟进写",
    "platform.shop.merchant.tag": "挂接已有标签",
    "platform.shop.merchant.tag.manage": "新建标签名",
    "platform.shop.moderate": "违规稽查",
    "platform.shop.onboarding.initiate": "帮客户发起入驻",
    "platform.shop.plan.manage": "套餐配置",
    "platform.shop.product.force_off": "商品强制下架",
    "platform.shop.product.review": "商品人审",
    "platform.shop.settlement": "清结算写",
    "platform.shop.subscription.manage": "开通/续费",
    "platform.shop.subscription.read": "查看套餐/到期",
    "platform.user.manage": "管理其他运营账号",
}

# P08-A 矩阵展示顺序（含默认未授予项，对照 06#p08a）
PLATFORM_SHOP_ROLE_MATRIX_CODES: dict[str, tuple[str, ...]] = {
    PLATFORM_SHOP_ROLE_CS: (
        "platform.shop.merchant.list_assigned",
        "platform.shop.merchant.read",
        "platform.shop.merchant.tag",
        "platform.shop.merchant.tag.manage",
        "platform.shop.onboarding.initiate",
        "platform.shop.subscription.read",
        "platform.shop.analytics",
        "platform.shop.approve",
        "platform.shop.subscription.manage",
        "platform.shop.merchant.manage",
        "platform.shop.merchant.assign",
    ),
    PLATFORM_SHOP_ROLE_OPS: (
        "platform.shop.merchant.list_all",
        "platform.shop.merchant.assign",
        "platform.shop.merchant.read",
        "platform.shop.merchant.tag",
        "platform.shop.merchant.tag.manage",
        "platform.shop.onboarding.initiate",
        "platform.shop.approve",
        "platform.shop.merchant.manage",
        "platform.shop.plan.manage",
        "platform.shop.subscription.manage",
        "platform.shop.product.review",
        "platform.shop.moderate",
        "platform.shop.product.force_off",
        "platform.shop.analytics",
        "platform.shop.settlement",
        "platform.user.manage",
    ),
    PLATFORM_SHOP_ROLE_FINANCE: (
        "platform.shop.settlement",
        "platform.shop.fee.manage",
        "platform.shop.subscription.read",
        "platform.shop.analytics",
        "platform.shop.approve",
        "platform.shop.subscription.manage",
        "platform.shop.merchant.manage",
    ),
}

PLATFORM_SHOP_ROLE_META: tuple[dict[str, str], ...] = (
    {
        "code": "",
        "name": "平台超管",
        "code_label": "platform_admin",
        "summary": "全部 platform.shop.* + 主站租户/账号管理",
    },
    {
        "code": PLATFORM_SHOP_ROLE_OPS,
        "name": "日常运营",
        "code_label": PLATFORM_SHOP_ROLE_OPS,
        "summary": "入驻/审核/暂停恢复、套餐开通、稽查；list_all 全站商家",
    },
    {
        "code": PLATFORM_SHOP_ROLE_CS,
        "name": "商家管家",
        "code_label": PLATFORM_SHOP_ROLE_CS,
        "summary": "仅 list_assigned 所辖客户 · 代建入驻 · 续费跟进 · 无审核/暂停/开通写",
    },
    {
        "code": PLATFORM_SHOP_ROLE_FINANCE,
        "name": "财务结算",
        "code_label": PLATFORM_SHOP_ROLE_FINANCE,
        "summary": "清结算、对账、费率只读、看板只读",
    },
)

PLATFORM_ADMIN_ROLE = "platform_admin"
